### PR TITLE
seL4 and linux kernel tracing supporting utils

### PR DESCRIPTION
**Pull request by jsvapiav**
* Add sel4-extract -- sel4 tracebuffer -> ftrace conversion tool
* Add sel4-ftrace-merge -- ftrace merge tool

jira: HYPR-432
Signed-off-by: Ivan Kuznetsov <jsvapiav@gmail.com>

### CODE REVIEW

File: README.md
Comment (elmankku) [inline README.md]: Put things after `-` to normal text.
```suggestion
# sel4-trace-support

A collection of programs to convert and merge seL4 tracebuffer to/with ftrace logs (WIP).
```

File: README.md
Comment (elmankku) [inline README.md:5]: It is generally recommended to have blank lines before and after a title.

```suggestion
### Build

```

File: README.md
Comment (elmankku) [inline README.md]: A typo fix, blank line etc.
```suggestion
#### sel4-extract

A tool for converting seL4 tracebuffer logs to ftrace compatible binary format.


```

File: README.md
Comment (elmankku) [inline README.md]: Fix typos and new line etc.

```suggestion
#### sel4-ftrace-merge

A tool for merging converted seL4 tracebuffer logs with trace-cmd\kernelshark logs.

```

File: README.md
Comment (elmankku) [inline README.md:33]: Newline
```suggestion
### Format description

```

File: README.md
Comment (elmankku) [inline README.md]: You could use [reference style links](https://www.markdownguide.org/basic-syntax/#reference-style-links)

```suggestion
trace-cmd binary format is described at `trace-cmd.dat` [man page][1] .

... text ...
... plaa ...

[1]: https://man7.org/linux/man-pages/man5/trace-cmd.dat.5.html
```

File: README.md
Comment (elmankku) [inline README.md]:
```suggestion
ftrace binary described in following parts of trace-cmd.dat (**_Please notice: currently formats are hardcoded, but they should be parsed from following places_**):
```

File: README.md
Comment (elmankku) [inline README.md]: Perhaps

```suggestion
+ HEADER INFO FORMAT - contains a page (4096b) header format and event header formats. Event header format can also be found `debugfs/tracing/events/header_event`-file
```

File: README.md
Comment (elmankku) [inline README.md]:
```suggestion
   Field descriptions:
```

File: README.md
Comment (elmankku) [inline README.md]:
```suggestion
      commit    -- number of bytes used in the page (up to 4080)
```

File: README.md
Comment (elmankku) [inline README.md]:
```suggestion
      data      -- data content of the page
```

File: README.md
Comment (elmankku) [inline README.md]:
```suggestion
   Field description:
```

File: README.md
Comment (elmankku) [inline README.md:75]: differs from above... intentional?

File: README.md
Comment (elmankku) [inline README.md]: Leading and trailing newline for clarity.
```suggestion

   Field descriptions:
   
```

File: README.md
Comment (elmankku) [inline README.md:167]: As mentioned above, some of these could be reference style links, and some could be like they currently are. Your call :)

**Review by elmankku — COMMENTED** [review]
Some comments for markdown.

File: README.md
Comment (jsvapiav) [reply README.md:75]: Sorry, I probably did not get you, but:
```
time_delta -- time difference with timestamp in PAGE HEADER (for 1st entry is zero)
```
time_delta is difference btwn timestamp in PAGE HEADER and time in current event. Is this sounds like an answer for your question?

File: README.md
Comment (jsvapiav) [reply README.md:167]: Sure, thank you for all your suggestions. But here I think that there is not much difference to use this approach or any other one. So keep them as is.

File: sel4-extract/sel4-extract.c
Comment (hlyytine) [inline sel4-extract/sel4-extract.c]: Do you have strong reason to stay with this Camel_Case capitalization? If not, could you just follow like the one used in Linux kernel, like `struct entry_header` and `struct event_function`.

File: sel4-ftrace-merge/ftrace-merge.c
Comment (hlyytine) [inline sel4-ftrace-merge/ftrace-merge.c]: Not sure if indentation is OK here.

File: sel4-ftrace-merge/ftrace-merge.c
Comment (hlyytine) [inline sel4-ftrace-merge/ftrace-merge.c]: Indentation?

File: sel4-ftrace-merge/ftrace-merge.c
Comment (hlyytine) [inline sel4-ftrace-merge/ftrace-merge.c]: Indentation?

File: sel4-ftrace-merge/ftrace-merge.c
Comment (hlyytine) [inline sel4-ftrace-merge/ftrace-merge.c:1]: Perhaps this entire file could benefit if checked with Linux's checkpatch.pl or some sort of autoindentation software?

File: sel4-ftrace-merge/include/ftrace.h
Comment (hlyytine) [inline sel4-ftrace-merge/include/ftrace.h:1]: Not sure if this file is free from indentation issues either.

**Review by hlyytine — CHANGES_REQUESTED** [review]
I'm happy with the code but style issues should be fixed before merging.

File: sel4-extract/sel4-extract.c
Comment (jsvapiav) [reply sel4-extract/sel4-extract.c]: Sorry, I was really thinking that I fixed that already. To many different repos are opened simultaneous on my laptop. Probably it time to keep vim and adopt some proper IDE.

File: README.md
Comment (elmankku) [reply README.md:75]: Nevermind :)

File: README.md
Comment (elmankku) [reply README.md:167]: Alright, np

File: sel4-ftrace-merge/include/ftrace.h
Comment (jsvapiav) [reply sel4-ftrace-merge/include/ftrace.h:1]: here again. I already fixed all thees  indentation issues, but submitted them on different repo. Sorry.

File: sel4-extract/CMakeLists.txt
Comment (elmankku) [inline sel4-extract/CMakeLists.txt]: I recommend using a configuration flag (i.e. what you pass to `cmake -DSOMETHING=1`) instead of comment and conditionally add either directly to target_compile_options or to a variable.

File: sel4-extract/CMakeLists.txt
Comment (elmankku) [inline sel4-extract/CMakeLists.txt]: Do not use handwritten paths, use [GNUInstallDirs](https://cmake.org/cmake/help/latest/module/GNUInstallDirs.html#module:GNUInstallDirs). That way it is guaranteed that the paths are being handled correctly (i.e. prefixes etc). This way cross compilation and integration to build systems is easier.

So use:
 ```suggestion
include(GNUInstallDirs)
 
install(TARGETS sel4-extract RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR})
```

File: sel4-ftrace-merge/CMakeLists.txt
Comment (elmankku) [inline sel4-ftrace-merge/CMakeLists.txt]: same comment apply here as with extract. In fact you could define these flags in common CMakeLists.txt (i.e. one dir above.

File: sel4-ftrace-merge/CMakeLists.txt
Comment (elmankku) [inline sel4-ftrace-merge/CMakeLists.txt]: Treat compiler and linker options separately, don't mush them up (i.e. own variables for preprocessor, compiler and linker flags. By convention these are called CPPFLAGS, CFLAGS and LDFLAGS respectively.

File: sel4-ftrace-merge/CMakeLists.txt
Comment (elmankku) [inline sel4-ftrace-merge/CMakeLists.txt]: Use GNUInstallDirs.

File: sel4-extract/CMakeLists.txt
Comment (elmankku) [inline sel4-extract/CMakeLists.txt]: Treat compiler and linker options separately, don't mush them up (i.e. own variables for preprocessor, compiler and linker flags. By convention these are called CPPFLAGS, CFLAGS and LDFLAGS respectively.

**Review by elmankku — CHANGES_REQUESTED** [review]

File: sel4-ftrace-merge/include/ftrace.h
Comment (hlyytine) [inline sel4-ftrace-merge/include/ftrace.h]: That opening brace belongs on the same line as `do`. The only exception (that comes to my mind right now) is function definition where the brace is on its own line.

File: sel4-ftrace-merge/include/ftrace.h
Comment (hlyytine) [inline sel4-ftrace-merge/include/ftrace.h]: Indent level here seems excessive? Please check the structs below as well, there seems to be problems with those.

File: sel4-ftrace-merge/include/ftrace.h
Comment (hlyytine) [inline sel4-ftrace-merge/include/ftrace.h:129]: I mean, this has too much indent?

File: sel4-ftrace-merge/include/ftrace.h
Comment (hlyytine) [inline sel4-ftrace-merge/include/ftrace.h]: I know we are not inside seL4 here, but seL4 coding guidelines dictate` #pragma once` should be used instead. The reason I'm bringing this up, because I know some standards dictate the naming scheme of the preprocessor identifier used. I would think nowadays compilers support `#pragma once` and this way we would not need to bother ourselves what's the correct naming. Opinions @elmankku ?

File: sel4-ftrace-merge/include/ftrace.h
Comment (hlyytine) [inline sel4-ftrace-merge/include/ftrace.h]: Also why I started my rant about `#pragma once` (or lack of it) is that I see this closing #endif sometimes prefixed with old style C comment and sometimes with C++ style comment like this. Outside Linux kernel both work with default compiler C dialect (C11), but I think we should strive for consistency, choose one style and stick to it. I was thinking `#pragma once` would be easiest solution, no need to decide anything.

File: sel4-ftrace-merge/ftrace-merge.c
Comment (hlyytine) [inline sel4-ftrace-merge/ftrace-merge.c]: To me indents here are wrong?

File: sel4-ftrace-merge/ftrace-merge.c
Comment (hlyytine) [inline sel4-ftrace-merge/ftrace-merge.c]: Add comma even if it's the last member, this way if you introduce new members to structs, you can define them here and still the patch would not need to include the addition of comma here.

**Review by hlyytine — CHANGES_REQUESTED** [review]

File: sel4-extract/CMakeLists.txt
Comment (jsvapiav) [reply sel4-extract/CMakeLists.txt]: Fixed.

File: sel4-extract/CMakeLists.txt
Comment (jsvapiav) [reply sel4-extract/CMakeLists.txt]: Fixed.

File: sel4-ftrace-merge/CMakeLists.txt
Comment (jsvapiav) [reply sel4-ftrace-merge/CMakeLists.txt]: Fixed.

File: sel4-ftrace-merge/CMakeLists.txt
Comment (jsvapiav) [reply sel4-ftrace-merge/CMakeLists.txt]: Fixed.

File: sel4-extract/CMakeLists.txt
Comment (jsvapiav) [reply sel4-extract/CMakeLists.txt]: Fixed.

File: sel4-ftrace-merge/include/ftrace.h
Comment (jsvapiav) [reply sel4-ftrace-merge/include/ftrace.h]: Sure, moved on the same line.

File: sel4-ftrace-merge/include/ftrace.h
Comment (jsvapiav) [reply sel4-ftrace-merge/include/ftrace.h:129]: Fixed.

File: sel4-ftrace-merge/ftrace-merge.c
Comment (jsvapiav) [reply sel4-ftrace-merge/ftrace-merge.c]: Fixed.

File: sel4-ftrace-merge/ftrace-merge.c
Comment (jsvapiav) [reply sel4-ftrace-merge/ftrace-merge.c]: Fixed.

File: sel4-ftrace-merge/include/ftrace.h
Comment (jsvapiav) [reply sel4-ftrace-merge/include/ftrace.h]: Fixed comment to 'C dialect'.

File: sel4-ftrace-merge/include/ftrace.h
Comment (jsvapiav) [reply sel4-ftrace-merge/include/ftrace.h]: Fixed.

File: sel4-ftrace-merge/include/ftrace.h
Comment (jsvapiav) [reply sel4-ftrace-merge/include/ftrace.h]: Me personally prefer to use #IFNDEF #ENDIF with a pure C, but let's check other suggestions.

File: sel4-ftrace-merge/CMakeLists.txt
Comment (jsvapiav) [reply sel4-ftrace-merge/CMakeLists.txt]: Fixed. Yeah!

File: sel4-ftrace-merge/include/ftrace.h
Comment (elmankku) [reply sel4-ftrace-merge/include/ftrace.h]: > I know we are not inside seL4 here, but seL4 coding guidelines dictate` #pragma once` should be used instead. The reason I'm bringing this up, because I know some standards dictate the naming scheme of the preprocessor identifier used. I would think nowadays compilers support `#pragma once` and this way we would not need to bother ourselves what's the correct naming. Opinions @elmankku ?

Yes agreed, `#pragma once` is good default for the new stuff - makes life a bit easier. Naturally we should follow repo/upstream conventions when applicable.

Comment (hlyytine) [reply sel4-ftrace-merge/include/ftrace.h]: What do you think @jsvapiav ? Could you live with `#pragma once`?

File: sel4-ftrace-merge/include/ftrace.h
Comment (hlyytine) [inline sel4-ftrace-merge/include/ftrace.h:25]: Sorry my bad, just checked how Linux sources do this: move `do {` to its own line, with indentation of one tab. Then it should be perfect.

File: sel4-extract/sel4-extract.c
Comment (hlyytine) [inline sel4-extract/sel4-extract.c]: I would like some consistency here: above you have bitfields aligned so that colon is in a definite column, followed by a space before the field length. Here the column is rather arbitrary and no space after colon. Please choose one and stick to that, I don't know which one is mandated by Linux kernel (which I regard the ultimate coding guide even in other projects), but personally I find the former more pleasing.

File: sel4-extract/sel4-extract.c
Comment (hlyytine) [inline sel4-extract/sel4-extract.c]: One comment: seems that whenever you have a line continuation, you seem to use more than one tab for indent. There are two schools on this: one is to use one tab only (I think this is what K&R suggest but I might be wrong). The second used in Linux kernel is that use as many tabs as you can and then fill up with spaces to get even with the first character after the parenthesis on the above line.

Comment (hlyytine) [reply sel4-extract/sel4-extract.c]: Seems that elsewhere you follow the former method which is OK to me and which I use unless I specifically try to write 100% Linux kernel style compatible code.

**Review by hlyytine — CHANGES_REQUESTED** [review]

File: sel4-ftrace-merge/include/ftrace.h
Comment (jsvapiav) [reply sel4-ftrace-merge/include/ftrace.h]: Sure, NP =)

File: sel4-ftrace-merge/include/ftrace.h
Comment (jsvapiav) [reply sel4-ftrace-merge/include/ftrace.h:25]: Actually, I did grep `/src/linux` and can find pretty different cases... And pretty much with `do {` on the same line. But, I'm fine with any =)

File: sel4-extract/sel4-extract.c
Comment (jsvapiav) [reply sel4-extract/sel4-extract.c]: This one is default vim behavior.. most time I'm manually fix that, but sometimes forget =)

File: sel4-extract/sel4-extract.c
Comment (jsvapiav) [reply sel4-extract/sel4-extract.c]: Sure, my bad -- fixed.

**Review by hlyytine — APPROVED** [review]

**Pull request merged**
